File: solutions/kotlin/crypto-square/1/src/main/kotlin/CryptoSquare.kt
import kotlin.math.floor
import kotlin.math.sqrt

object CryptoSquare {

    fun ciphertext(plaintext: String): String {
        val plaintext = plaintext.filter{ it.isLetterOrDigit() }.lowercase()
        val size = plaintext.length
        var r = floor(sqrt(size.toDouble())).toInt()
        var c = r
        var change_r = false
        while (r * c < size) {
            if (change_r) {
                r++
            } else {
                c++
            }
            change_r = !change_r
        }
        return (0..<c)
            .fold(StringBuilder()) { builder, ci ->
                (0..<r)
                    .fold(builder) { builder, ri ->
                        builder.append(plaintext.getOrElse(ri*c+ci) { ' ' })
                        builder
                    }
                    .also {
                        if (ci + 1 != c) {
                            builder.append(' ')   
                        }
                    }
            }
            .toString()
    }

}

File: solutions/typescript/crypto-square/1/crypto-square.ts
export class Crypto {
  private cipher: string
  
  constructor(plainText: string) {
    const sanitized = Array.from(plainText.toLowerCase())
      .filter(c => ('a' <= c && c <= 'z') || ('0' <= c && c <= '9'))
    let r: number
    let c: number
    const size = sanitized.length
    const root = Math.floor(Math.sqrt(size))
    if (root * root === size) {
      [r, c] = [root, root]
    } else if (root * (root + 1) >= size) {
      [r, c] = [root, root + 1]
    } else {
      [r, c] = [root + 1, root + 1]
    }
    const result = []
    for (let ci = 0; ci < c; ++ci) {
      for (let ri = 0; ri < r; ++ri) {
        const textIdx = ri * c + ci
        if (textIdx < size) {
          result.push(sanitized[textIdx])
        } else {
          result.push(' ')
        }
      }
      if (ci != c - 1) {
        result.push(' ')
      }
    }
    this.cipher = result.join('')
  }

  get ciphertext(): string {
    return this.cipher
  }
}